src/pip/_internal/index.py

```diff
@@ -1333,9 +1336,17 @@ def _clean_link(url):
         path = urllib_request.pathname2url(
             urllib_request.url2pathname(result.path))
     else:
-        # In addition to the `/` character we protect `@` so that
-        # revision strings in VCS URLs are properly parsed.
-        path = urllib_parse.quote(urllib_parse.unquote(result.path), safe="/@")
+        # In addition to the `/` character we protect `@`
+        # (as well as theirs %-escapes) so that revision
+        # strings in VCS URLs are properly parsed.
+        path_bits = _safe_chars_re.split(result.path)
+        path_parts = [urllib_parse.quote(urllib_parse.unquote(path_bits[0]))]
+        for i in range(1, len(path_bits), 2):
+            path_parts.append(path_bits[i])
+            path_parts.append(
+                urllib_parse.quote(urllib_parse.unquote(path_bits[i + 1]))
+            )
+        path = ''.join(path_parts)
```

Comment:
Sorry for jumping in on this later. I think the code inside this `else` block should definitely be split out into an independently tested function, since it's a bit tricky. (I had a partially finished PR for this issue as well on my machine, using a similar approach. I'd also like to compare with what I have to see if there is anything else I noticed.)

Reply:
Actually, here is the PR I started working on a while ago. It's related but not quite the same: https://github.com/pypa/pip/pull/6496

Reply:
> I think the code inside this `else` block should definitely be split out into an independently tested function, since it's a bit tricky.

That sounds reasonable.

> Actually, here is the PR I started working on a while ago. It's related but not quite the same: #6496

I see you have almost finished it. What should we do with the current PR?

Reply:
> I see you have almost finished it. What should we do with the current PR?

I would just examine the other PR for any differences to see if anything might be missing here and/or worth carrying over. Like, I noticed the other PR also extends the treatment to the `file://` case, which I haven't thought recently about to know if it's desired. I think the test cases in the other PR would be something else worth carrying over (at least the applicable ones) since I had put some thought into those.

Reply:
Also, I think the "revision" portion of the other PR (e.g. the changes to `make_vcs_requirement_url()`) can be left for that issue / PR, as that issue is distinct from the one this PR is trying to resolve.

Reply:
May i cherry-pick 2133c1d0133dc0eb23bd6cbea6d012c1cf711d2d and b978ff6ce0cb2f4011b2666072a5e5a00bd7c26b?

Reply:
Sure, whatever is easiest for you.. You'll probably want to be making changes on top of that though as I'm guessing you will. For example, I didn't use the regex approach as I was splitting only on one character for that PR.

Reply:
> You'll probably want to be making changes on top of that though as I'm guessing you will

Thanks! That's exactly i was going to do :)